Quotient/connectionencryptiondata_p.cpp

```diff
@@ -18,102 +18,109 @@
-        WritePasswordJob writeJob(qAppName());
-        writeJob.setAutoDelete(false);
-        writeJob.setKey(keychainId);
-        writeJob.setBinaryData(picklingKey.viewAsByteArray());
-        QEventLoop writeLoop;
-        QObject::connect(&writeJob, &Job::finished, &writeLoop,
-                         &QEventLoop::quit);
-        writeJob.start();
-        writeLoop.exec();
-
-        if (writeJob.error() == Error::NoError)
-            return std::move(picklingKey);
-
-        qCritical(E2EE) << "Could not save pickling key to keychain: "
-                        << writeJob.errorString();
-        return writeJob.error();
-    }
-    qWarning(E2EE) << "Error loading pickling key - please fix your keychain:"
-                   << readJob.errorString();
-    return readJob.error();
-}
-
-std::optional<std::unique_ptr<ConnectionEncryptionData>>
-ConnectionEncryptionData::setup(Connection* connection, bool mock)
-{
-    if (auto&& maybePicklingKey = setupPicklingKey(connection->userId(), mock)) {
-        auto&& encryptionData = std::make_unique<ConnectionEncryptionData>(
-            connection, std::move(*maybePicklingKey));
-        if (mock) {
-            encryptionData->database.clear();
-            encryptionData->olmAccount.setupNewAccount();
-            return std::move(encryptionData);
+        if (readJob->error() == Error::EntryNotFound) {
+            auto&& picklingKey = PicklingKey::generate();
+            auto writeJob = new WritePasswordJob(qAppName());
+            writeJob->setAutoDelete(true);
+            writeJob->setKey(keychainId);
+            writeJob->setBinaryData(picklingKey.viewAsByteArray());
+            QObject::connect(writeJob, &Job::finished, writeJob, [writeJob, then, &picklingKey](){
+                writeJob->start();
+
+                if (writeJob->error() == Error::NoError) {
+                    then(std::move(picklingKey));
+                    return;
+                }
+
+                qCritical(E2EE) << "Could not save pickling key to keychain: "
+                                << writeJob->errorString();
+                then(writeJob->error());
+            });
         }
-        if (const auto outcome = encryptionData->database.setupOlmAccount(
-                encryptionData->olmAccount)) {
-            // account already existing or there's an error unpickling it
-            if (outcome == OLM_SUCCESS)
-                return std::move(encryptionData);
-
-            qCritical(E2EE) << "Could not unpickle Olm account for"
-                            << connection->objectName();
-            return {};
+        qWarning(E2EE) << "Error loading pickling key - please fix your keychain:"
+                    << readJob->errorString();
+        then(readJob->error());
+        return;
+    });
+    readJob->start();
+}
+
+void ConnectionEncryptionData::setup(Connection* connection, std::function<void(std::optional<std::unique_ptr<ConnectionEncryptionData>>)> then, bool mock)
```

Comment:
...and then this function can be something like `return setupPicklingKey(...).then([] (Expected<...>) /*-> std::optional<...> */ { ... });` - or you can even reuse the `QFuture`'s cancelled state and strip `std::optional`, instead cancelling the future when the setup didn't work out. Again, I understand it's a change of paradigm so I can do it if you are not sure.